README.md
# Zapier vs. Fitbit

Fitbit integration for Zapier platform.
Not much going on for the time being, but still, it's a fun one for me!

## Scope

The scope is currently limited to `weight` (weight, body mass index, body fat percentage, and goals) and `nutrition` (calorie consumption, food/water logging, goals, and plans). Next ones up would be `profile`, `sleep`, `heartrate`, and `activity`.

## Triggers

[x] Body Fat (_weight_ scope)
[x] Body Weight (_weight_ scope)

### Up next…

[ ] Water consumption (_nutrition_ scope)
[ ] Calories in (_nutrition_ scope)
[ ] Nutrition facts (_nutrition_ scope)

## Development notes

1. `npm i` to install dependencies
2. `npm t` to run the tests


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.4.5"></a>
## [0.4.5](https://github.com/jsDotCr/zapier-fitbit/compare/v0.4.4...v0.4.5) (2018-06-02)



<a name="0.4.4"></a>
## [0.4.4](https://github.com/jsDotCr/zapier-fitbit/compare/v0.4.3...v0.4.4) (2018-03-05)


### Bug Fixes

* **authentication:** store refresh token as well as the access token ([a02f537](https://github.com/jsDotCr/zapier-fitbit/commit/a02f537))



<a name="0.4.3"></a>
## [0.4.3](https://github.com/jsDotCr/zapier-fitbit/compare/v0.4.2...v0.4.3) (2018-03-04)


### Bug Fixes

* **auth:** missing return ([72c7eb1](https://github.com/jsDotCr/zapier-fitbit/commit/72c7eb1))



<a name="0.4.2"></a>
## [0.4.2](https://github.com/jsDotCr/zapier-fitbit/compare/v0.4.1...v0.4.2) (2018-03-04)


### Bug Fixes

* **auth:** do not override Authentication header when it's set with Basic ([014f6bb](https://github.com/jsDotCr/zapier-fitbit/commit/014f6bb))



<a name="0.4.1"></a>
## [0.4.1](https://github.com/jsDotCr/zapier-fitbit/compare/v0.4.0...v0.4.1) (2018-03-03)


### Bug Fixes

* **body:** simplify id ([9a33e07](https://github.com/jsDotCr/zapier-fitbit/commit/9a33e07))
* **body weight:** update test with more realistic data ([0e1aa18](https://github.com/jsDotCr/zapier-fitbit/commit/0e1aa18))
* **body weight:** update test with more realistic data ([66b365e](https://github.com/jsDotCr/zapier-fitbit/commit/66b365e))



<a name="0.4.0"></a>
# 0.4.0 (2018-03-03)


### Bug Fixes

* **auth:** fix test endpoint ([3c56b00](https://github.com/jsDotCr/zapier-fitbit/commit/3c56b00))
* **authentication:** fix wrong variable name ([24aab38](https://github.com/jsDotCr/zapier-fitbit/commit/24aab38))
* **authentication:** missing redirect_url, update refresh token payload ([bf2e60e](https://github.com/jsDotCr/zapier-fitbit/commit/bf2e60e))
* **body:** create a unique ID ([343a3df](https://github.com/jsDotCr/zapier-fitbit/commit/343a3df))
* **body:** fix base date ([8bc7fc7](https://github.com/jsDotCr/zapier-fitbit/commit/8bc7fc7))
* **body:** reverse array of items ([5bfba75](https://github.com/jsDotCr/zapier-fitbit/commit/5bfba75))
* **body:** reverse array of items ([fdbadcb](https://github.com/jsDotCr/zapier-fitbit/commit/fdbadcb))
* **test:** update references to app and list of tests to execute ([3adf1ec](https://github.com/jsDotCr/zapier-fitbit/commit/3adf1ec))


### Features

* **authentication:** add authentication functions ([24ca55d](https://github.com/jsDotCr/zapier-fitbit/commit/24ca55d))
* **body fat:** add bodyFat measurement trigger ([008a7e0](https://github.com/jsDotCr/zapier-fitbit/commit/008a7e0))
* **body weight:** add body weight trigger ([c99a40e](https://github.com/jsDotCr/zapier-fitbit/commit/c99a40e))
* **env:** publish env variables from .environment file ([b4914e8](https://github.com/jsDotCr/zapier-fitbit/commit/b4914e8))



<a name="0.3.4"></a>
## [0.3.4](https://github.com/jsDotCr/zapier-fitbit/compare/v0.3.3...v0.3.4) (2018-03-03)


### Bug Fixes

* **body:** reverse array of items ([2d3b6fb](https://github.com/jsDotCr/zapier-fitbit/commit/2d3b6fb))
* **body:** reverse array of items ([d6f51a1](https://github.com/jsDotCr/zapier-fitbit/commit/d6f51a1))
* **test:** update references to app and list of tests to execute ([d22057a](https://github.com/jsDotCr/zapier-fitbit/commit/d22057a))



<a name="0.3.3"></a>
## [0.3.3](https://github.com/jsDotCr/zapier-fitbit/compare/v0.3.2...v0.3.3) (2018-03-03)


### Bug Fixes

* **auth:** fix test endpoint ([eb3e834](https://github.com/jsDotCr/zapier-fitbit/commit/eb3e834))
* **authentication:** fix wrong variable name ([b371514](https://github.com/jsDotCr/zapier-fitbit/commit/b371514))



<a name="0.3.2"></a>
## [0.3.2](https://github.com/jsDotCr/zapier-fitbit/compare/v0.3.1...v0.3.2) (2018-02-17)


### Bug Fixes

* **body:** fix base date ([5d444c2](https://github.com/jsDotCr/zapier-fitbit/commit/5d444c2))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/jsDotCr/zapier-fitbit/compare/v0.3.0...v0.3.1) (2018-02-16)



<a name="0.3.0"></a>
# 0.3.0 (2018-02-16)


### Bug Fixes

* **authentication:** missing redirect_url, update refresh token payload ([9eff1d9](https://github.com/jsDotCr/zapier-fitbit/commit/9eff1d9))
* **body:** create a unique ID ([978de9d](https://github.com/jsDotCr/zapier-fitbit/commit/978de9d))


### Features

* **authentication:** add authentication functions ([123e197](https://github.com/jsDotCr/zapier-fitbit/commit/123e197))
* **body fat:** add bodyFat measurement trigger ([84c06aa](https://github.com/jsDotCr/zapier-fitbit/commit/84c06aa))
* **body weight:** add body weight trigger ([9240c5e](https://github.com/jsDotCr/zapier-fitbit/commit/9240c5e))


test/triggers/bodyFat.js
/* globals describe it */
const moment = require('moment')
const should = require('should')
const zapier = require('zapier-platform-core')
const nock = require('nock')

const App = require('../../index')
const appTester = zapier.createAppTester(App)
zapier.tools.env.inject()

describe('Fitbit App', () => {
  it('should fetch last month\'s data', done => {
    const today = moment().format('YYYY-MM-DD')
    const oneMonthAgo = moment().subtract(1, 'month').format('YYYY-MM-DD')
    const bundle = { inputData: {} }
    nock('https://api.fitbit.com/1')
      .replyContentLength()
      .get(new RegExp(`/user/-/body/log/fat/date/${oneMonthAgo}/${today}.json`))
      .query(bundle.inputData)
      .reply(200, {
        'fat': [
          {
            'date': '2012-03-05',
            'fat': 14,
            'logId': 1330991999000,
            'time': '23:59:59',
            'source': 'API'
          }
        ]
      })

    appTester(App.triggers.bodyFat.operation.perform, bundle)
      .then(results => {
        should.exist(results)
        done()
      })
      .catch(done)
  })

  it('should run triggers.bodyFat', done => {
    const bundle = { inputData: {} }
    nock('https://api.fitbit.com/1')
      .replyContentLength()
      .get(/\/user\/-\/body\/log\/fat\/date.*\.json/)
      .query(bundle.inputData)
      .reply(200, {
        'fat': [
          {
            'date': '2012-03-05',
            'fat': 14,
            'logId': 1330991999000,
            'time': '23:59:59',
            'source': 'API'
          },
          {
            'date': '2012-03-05',
            'fat': 13.5,
            'logId': 1330991999000,
            'time': '21:20:59',
            'source': 'Aria'
          }
        ]
      }
      )

    appTester(App.triggers.bodyFat.operation.perform, bundle)
      .then(results => {
        results.length.should.equal(2)
        done()
      })
      .catch(done)
  })
})


triggers/bodyFat.test.js
/* globals describe it */
const moment = require('moment')
const should = require('should')
const zapier = require('zapier-platform-core')
const nock = require('nock')

const App = require('../index')
const appTester = zapier.createAppTester(App)
zapier.tools.env.inject()

describe('Fitbit App triggers: body fat', () => {
  it('should fetch last month\'s data', done => {
    const today = moment().format('YYYY-MM-DD')
    const oneMonthAgo = moment().subtract(1, 'month').format('YYYY-MM-DD')
    const bundle = { inputData: {} }
    nock('https://api.fitbit.com/1')
      .replyContentLength()
      .get(new RegExp(`/user/-/body/log/fat/date/${oneMonthAgo}/${today}.json`))
      .query(bundle.inputData)
      .reply(200, {
        'fat': [
          {
            'date': '2012-03-05',
            'fat': 14,
            'logId': 1330991999000,
            'time': '23:59:59',
            'source': 'API'
          }
        ]
      })

    appTester(App.triggers.bodyFat.operation.perform, bundle)
      .then(results => {
        should.exist(results)
        done()
      })
      .catch(done)
  })

  it('should run triggers.bodyFat', done => {
    const bundle = { inputData: {} }
    nock('https://api.fitbit.com/1')
      .replyContentLength()
      .get(/\/user\/-\/body\/log\/fat\/date.*\.json/)
      .query(bundle.inputData)
      .reply(200, {
        'fat': [
          {
            'date': '2012-03-05',
            'fat': 14,
            'logId': 1330991799000,
            'time': '23:59:59',
            'source': 'API'
          },
          {
            'date': '2012-03-06',
            'fat': 13.5,
            'logId': 1330991899000,
            'time': '21:20:59',
            'source': 'Aria'
          },
          {
            'date': '2012-03-07',
            'fat': 13.8,
            'logId': 1330991999000,
            'time': '20:20:59',
            'source': 'Aria'
          }
        ]
      }
      )

    appTester(App.triggers.bodyFat.operation.perform, bundle)
      .then(results => {
        results.length.should.equal(3)
        done()
      })
      .catch(done)
  })

  it('should reverse the order of elements', done => {
    const bundle = { inputData: {} }
    const elements = [
      {
        'date': '2012-03-05',
        'fat': 14,
        'logId': 1330991799000,
        'time': '23:59:59',
        'source': 'API'
      },
      {
        'date': '2012-03-06',
        'fat': 13.5,
        'logId': 1330991899000,
        'time': '21:20:59',
        'source': 'Aria'
      },
      {
        'date': '2012-03-07',
        'fat': 13.8,
        'logId': 1330991999000,
        'time': '20:20:59',
        'source': 'Aria'
      }
    ]
    nock('https://api.fitbit.com/1')
      .replyContentLength()
      .get(/\/user\/-\/body\/log\/fat\/date.*\.json/)
      .query(bundle.inputData)
      .reply(200, {
        fat: elements
      })

    appTester(App.triggers.bodyFat.operation.perform, bundle)
      .then(results => {
        results[0].date.should.equal(elements[elements.length - 1].date)
        results[results.length - 1].date.should.eql(elements[0].date)
        done()
      })
      .catch(done)
  })
})


test/triggers/bodyWeight.js
/* globals describe it */
const should = require('should')
const zapier = require('zapier-platform-core')
const nock = require('nock')

const App = require('../../index')
const appTester = zapier.createAppTester(App)
zapier.tools.env.inject()

describe('Fitbit App', () => {
  it('should run triggers.bodyWeight', done => {
    const bundle = { inputData: {} }
    nock('https://api.fitbit.com/1')
      .replyContentLength()
      .get(/\/user\/-\/body\/log\/weight\/date.*\.json/)
      .query(bundle.inputData)
      .reply(200, {
        'weight': [
          {
            'bmi': 23.57,
            'date': '2015-03-05',
            'logId': 1330991999000,
            'time': '23:59:59',
            'weight': 73,
            'source': 'API'
          },
          {
            'bmi': 22.57,
            'date': '2015-03-05',
            'logId': 1330991999000,
            'time': '21:10:59',
            'weight': 72.5,
            'source': 'Aria'
          }
        ]
      })

    appTester(App.triggers.bodyWeight.operation.perform, bundle)
      .then(results => {
        should.exist(results)
        results.length.should.equal(2)
        done()
      })
      .catch(done)
  })
})


index.js
const authentication = require('./authentication')
const BodyFat = require('./triggers/bodyFat')
const BodyWeight = require('./triggers/bodyWeight')

const includeBearerToken = (request, z, bundle) => {
  if (request.headers.Authorization && request.headers.Authorization.startsWith('Basic')) {
    return request
  }
  if (bundle.authData.access_token) {
    request.headers.Authorization = `Bearer ${bundle.authData.access_token}`
  }
  return request
}

const App = {
  version: require('./package.json').version,
  platformVersion: require('zapier-platform-core').version,

  authentication: authentication,

  beforeRequest: [
    includeBearerToken
  ],

  afterResponse: [
  ],

  resources: {
  },

  triggers: {
    [BodyFat.key]: BodyFat,
    [BodyWeight.key]: BodyWeight
  },

  searches: {
  },

  creates: {
  }
}

module.exports = App


authentication.js
function getHeaders (clientId, clientSecret) {
  const headers = {
    'content-type': 'application/x-www-form-urlencoded'
  }
  if (clientId && clientSecret) {
    headers['Authorization'] = `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString('base64')}`
  }
  return headers
}

function handleTokensResponse (z, response) {
  if (response.status !== 200) {
    throw new Error(`Unable to fetch tokens. API responded with status ${response.status} saying "${response.content}"`)
  }

  const result = z.JSON.parse(response.content)
  return {
    access_token: result.access_token,
    refresh_token: result.refresh_token
  }
}

const getAccessToken = (z, bundle) => {
  const clientId = process.env.CLIENT_ID
  const clientSecret = process.env.CLIENT_SECRET
  const request = z.request('https://api.fitbit.com/oauth2/token', {
    method: 'POST',
    body: {
      code: bundle.inputData.code,
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: 'authorization_code',
      state: '{{bundle.inputData.state}}',
      redirect_uri: '{{bundle.inputData.redirect_uri}}'
    },
    headers: getHeaders(clientId, clientSecret)
  })

  return request.then(response => handleTokensResponse(z, response))
}

const refreshAccessToken = (z, bundle) => {
  const clientId = process.env.CLIENT_ID
  const clientSecret = process.env.CLIENT_SECRET
  const request = z.request('https://api.fitbit.com/oauth2/token', {
    method: 'POST',
    body: {
      grant_type: 'refresh_token',
      refresh_token: bundle.authData.refresh_token,
      expires_in: process.env.TOKEN_EXPIRE_TIME || 28800
    },
    headers: getHeaders(clientId, clientSecret)
  })

  return request.then(response => handleTokensResponse(z, response))
}

const testAuth = (z, bundle) => {
  const promise = z.request({
    method: 'POST',
    url: 'https://api.fitbit.com/oauth2/introspect',
    headers: getHeaders(),
    body: {
      token: bundle.authData.access_token
    }
  })

  return promise.then((response) => {
    if (response.status === 401) {
      throw new Error('The access token you supplied is not valid')
    }
    return response
  })
}

module.exports = {
  type: 'oauth2',
  oauth2Config: {
    authorizeUrl: {
      url: 'https://www.fitbit.com/oauth2/authorize',
      params: {
        client_id: '{{process.env.CLIENT_ID}}',
        state: '{{bundle.inputData.state}}',
        redirect_uri: '{{bundle.inputData.redirect_uri}}',
        response_type: 'code',
        scope: '{{process.env.SCOPE}}',
        expires_in: '2592000'
      }
    },
    getAccessToken: getAccessToken,
    refreshAccessToken: refreshAccessToken,
    autoRefresh: true
  },
  test: testAuth
}


triggers/bodyWeight.js
const { apiRequest } = require('./body')

module.exports = {
  key: 'bodyWeight',
  noun: 'Body Weight',
  display: {
    label: 'Get Body Weight',
    description: 'Triggers when a new body weight measurement is added.'
  },
  operation: {
    inputFields: [],
    perform: (z, bundle) => apiRequest('weight', z, bundle),
    sample: {
      id: '1330991999000-weight-2012-03-05T23:59:59',
      bmi: 23.57,
      date: '2015-03-05',
      logId: 1330991999000,
      time: '23:59:59',
      weight: 73,
      source: 'Aria'
    },
    outputFields: [
      {
        key: 'bmi',
        label: 'BMI'
      },
      {
        key: 'date',
        label: 'Date'
      },
      {
        key: 'logId',
        label: 'Log ID'
      },
      {
        key: 'time',
        label: 'Time'
      },
      {
        key: 'weight',
        label: 'Weight'
      },
      {
        key: 'source',
        label: 'Source'
      }
    ]
  }
}
